### PR TITLE
Migrate Razor SDK into repo

### PR DESCRIPTION
**Pull request by captainsafia**
**Changes in this PR**
- Added a few new test assets for the Razor SDK integration tests in `src/Assets/TestProjects`. All are direct copies of the test assets that exist in the aspnetcore repo, with the notable exception that they reference MVC packages directly instead of using a shim where needed.
- Removed Razor SDK from src/Layout/redist/targets/BundledSdks.targets.
- Add in-source Razor SDK to src/Layout/redist/targets/GenerateLayout.targets.
- Added Razor SDK targets under `RazorSdk/Targets` directory.
- Added Razor SDK tools under `RazorSdk/Tools` with locally scoped command line tools.
- Removed RazorSDK from Stage0SdkFiles item group in src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj.
- Added tests for Razor SDK tool under `src/Tests/Microsoft.NET.Sdk.Razor.Tool.Tests`
- Added unit and integration tests for Razor SDK under `src/Tests/Microsoft.NET.Sdk.Razor.Tests`

### CODE REVIEW

File: src/Assets/TestPackages/PackageLibraryDirectDependency/PackageLibraryTransitiveDependency/PackageLibraryTransitiveDependency.csproj [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    <Copyright>© Microsoft</Copyright>
+    <Product>Razor Test</Product>
+    <Company>Microsoft</Company>
+    <Description>PackageLibraryTransitiveDependency Description</Description>
+    <DeterministicSourcePaths>false</DeterministicSourcePaths>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <!-- We don't want to run build server when not running as tests. -->
+    <UseRazorBuildServer>false</UseRazorBuildServer>
```
Comment (pranavkm) [inline src/Assets/TestPackages/PackageLibraryDirectDependency/PackageLibraryTransitiveDependency/PackageLibraryTransitiveDependency.csproj:19]: Feel free to remove these..

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/TestFiles/Generated/Counter.razor.rz.scp.css [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.counter {
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tests/TestFiles/Generated/Counter.razor.rz.scp.css:1]: Are these used by tests?

File: src/RazorSdk/Tool/Microsoft.NET.Sdk.Razor.Tool.csproj [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    <!-- This is not a package, it is part of Razor SDK. -->
+    <IsPackable>false</IsPackable>
+    <IsShipping>false</IsShipping>
+
+    <!-- Don't produce rzc.exe -->
+    <UseAppHost>false</UseAppHost>
+    <RuntimeIdentifier />
+
+    <!-- Need to build this project in source build -->
+    <ExcludeFromSourceBuild>false</ExcludeFromSourceBuild>
```
Comment (pranavkm) [inline src/RazorSdk/Tool/Microsoft.NET.Sdk.Razor.Tool.csproj:19]: Is this a thing in this repo? I was under the impression `ExcludeFromSourceBuild` is an aspnetcore thing.

File: src/RazorSdk/Tasks/Microsoft.NET.Sdk.Razor.Tasks.csproj [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  <ItemGroup>
+    <AdditionalContent Include="$(RazorSdkRoot)Targets\**\*.*">
+      <Pack>true</Pack>
+      <PackagePath>targets</PackagePath>
+    </AdditionalContent>
+    <AdditionalContent Include="$(RazorSdkRoot)Targets\**\*.*">
+      <Pack>true</Pack>
+      <PackagePath>build\netstandard2.0</PackagePath>
+    </AdditionalContent>
+    <AdditionalContent Include="$(RazorSdkRoot)buildMultiTargeting\**\*.*">
```
Comment (pranavkm) [inline src/RazorSdk/Tasks/Microsoft.NET.Sdk.Razor.Tasks.csproj:59]: Feel free to remove this + the `build` directory that contain targets . We're no longer shipping this as a NuGet package so they're never used.

File: src/RazorSdk/Tasks/Microsoft.NET.Sdk.Razor.Tasks.csproj [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    <NoPackageAnalysis>true</NoPackageAnalysis>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
+    <!-- MSBuild Task DLLs need to be versioned with every build -->
+    <AutoGenerateAssemblyVersion>true</AutoGenerateAssemblyVersion>
+
+    <OutputPath>$(BaseOutputPath)</OutputPath>
+    <IsPackable>true</IsPackable>
```
Comment (pranavkm) [inline src/RazorSdk/Tasks/Microsoft.NET.Sdk.Razor.Tasks.csproj:26]: I don't think you need to pack these?

File: src/RazorSdk/Tasks/Microsoft.NET.Sdk.Razor.Tasks.csproj [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+  <PropertyGroup>
+    <Description>SDK for building and publishing Razor applications.</Description>
+    <OutputType>Library</OutputType>
+    <RootNamespace>Microsoft.NET.Sdk.Razor</RootNamespace>
+    <TargetFrameworks>$(SdkTargetFramework);net46</TargetFrameworks>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Build.Framework" Version="15.8.166" ExcludeAssets="Runtime" />
```
Comment (pranavkm) [inline src/RazorSdk/Tasks/Microsoft.NET.Sdk.Razor.Tasks.csproj:39]: Could we use use common versions here / target net472 ?

File: src/RazorSdk/Tasks/Microsoft.NET.Sdk.Razor.Tasks.csproj [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="15.8.166" ExcludeAssets="Runtime" />
+    <PackageReference Include="System.Reflection.Metadata"  Version="$(SystemReflectionMetadataVersion)" Condition="'$(TargetFramework)'=='net46'" />
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.Razor.Extensions"
+      Version="5.0.1"
```
Comment (pranavkm) [inline src/RazorSdk/Tasks/Microsoft.NET.Sdk.Razor.Tasks.csproj:43]: Could you change this to use $(...packageVersion) that gets updated via dependency flow. We don't want to hardcode this.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tool.Tests/TempDirectory.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright(c) .NET Foundation.All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+
+namespace Microsoft.NET.Sdk.Razor.Tool
+{
+    internal class TempDirectory : IDisposable
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tool.Tests/TempDirectory.cs:9]: Could we use `TestAssetManager.GetTempDirectory()` here?

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/MvcBuildIntegrationTest50.cs [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+using Microsoft.NET.TestFramework;
+using Microsoft.NET.TestFramework.Assertions;
+using Microsoft.NET.TestFramework.Commands;
+using Microsoft.NET.TestFramework.Utilities;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.NET.Sdk.Razor.Tests
+{
+    public class MvcBuildIntegrationTest50 : MvcBuildIntegrationTestLegacy
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tests/MvcBuildIntegrationTest50.cs:16]: nice

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs [hunk 10]
```diff
@@ -0,0 +1,856 @@
...
+
+        [Fact]
+        public void Build_SimpleMvc_UsingDotnetMSBuildAndWithoutBuildServer_CanBuildSuccessfully()
+            => Build_SimpleMvc_WithoutBuildServer_CanBuildSuccessfully();
+
+        [FullMSBuildOnlyFactAttribute]
+        public void Build_SimpleMvc_UsingDesktopMSBuildAndWithoutBuildServer_CanBuildSuccessfully()
+            => Build_SimpleMvc_WithoutBuildServer_CanBuildSuccessfully();
+
+        // This test is identical to the ones in BuildServerIntegrationTest except this one explicitly disables the Razor build server.
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs:34]: I'd remove this comment and remove `p:UseRazorBuildServer=false` from the test.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs [hunk 11]
```diff
@@ -0,0 +1,856 @@
...
+        private void Build_SimpleMvc_WithoutBuildServer_CanBuildSuccessfully()
+        {
+            var testAsset = "SimpleMvc";
+            var projectDirectory = _testAssetsManager
+                .CopyTestAsset(testAsset)
+                .WithSource();
+
+            var build = new BuildCommand(projectDirectory);
+            var outputPath = build.GetOutputDirectory("net5.0", "Debug").ToString();
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs:43]: This is a bit problematic. We usually always test against bleeding edge and use legacy tests to verify older projects continue to build. I don't want us to hardcode an exact tfm here. Do you have ideas on how to solve this?

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs [hunk 12]
```diff
@@ -0,0 +1,856 @@
...
+
+            var intermediateOutputPath = build.GetIntermediateDirectory("net5.0", "Debug").ToString();
+            var razorAssemblyInfo = Path.Combine(intermediateOutputPath, "SimpleMvc.RazorAssemblyInfo.cs");
+
+            new FileInfo(razorAssemblyInfo).Should().NotExist();
+            new FileInfo(Path.Combine(intermediateOutputPath, "SimpleMvc.RazorTargetAssemblyInfo.cs")).Should().NotExist();
+        }
+
+        [Fact]
+        public void Build_AddsRelatedAssemblyPart_IfCompileOnPublishIsAllowed()
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs:520]: Feel free to remove this test.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs [hunk 13]
```diff
@@ -0,0 +1,856 @@
...
+
+            var intermediateOutputPath = build.GetIntermediateDirectory("net5.0", "Debug").ToString();
+            var razorAssemblyInfo = Path.Combine(intermediateOutputPath, "SimpleMvc.RazorAssemblyInfo.cs");
+
+            new FileInfo(razorAssemblyInfo).Should().Exist();
+            new FileInfo(razorAssemblyInfo).Should().Contain("[assembly: Microsoft.AspNetCore.Mvc.ApplicationParts.RelatedAssemblyAttribute(\"SimpleMvc.Views\")]");
+        }
+
+        [Fact]
+        public void Build_WithGenerateRazorHostingAssemblyInfoFalse_DoesNotGenerateHostingAttributes()
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs:558]: Remove the hosting assembly info tests if you'd like. Nothing uses these attributes, so we don't super care about these any longer.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs [hunk 14]
```diff
@@ -0,0 +1,856 @@
...
+            var projectDirectory = _testAssetsManager
+                .CopyTestAsset(testAsset)
+                .WithSource();
+
+            var build = new BuildCommand(projectDirectory);
+            build.Execute("/p:AddRazorSupportForMvc=false").Should().HaveStdOutContaining("RAZORSDK1004");
+        }
+
+        [Fact]
+        public void Build_WithNoResolvedRazorConfiguration()
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs:829]: Feel free to remove. This is not a scenario anybody would run in to.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/TestFiles/Generated/Counter.razor.rz.scp.css (hunk 2 above)
Comment (captainsafia) [reply src/Tests/Microsoft.NET.Sdk.Razor.Tests/TestFiles/Generated/Counter.razor.rz.scp.css:1]: Yes, they are used by the concatenate file tests.

File: src/RazorSdk/Tasks/Microsoft.NET.Sdk.Razor.Tasks.csproj (hunk 5 above)
Comment (captainsafia) [reply src/RazorSdk/Tasks/Microsoft.NET.Sdk.Razor.Tasks.csproj:26]: I think it is needed based on the way the generate layout targets work. Both the Web SDK and Publish SDK have it set.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs (hunk 11 above)
Comment (captainsafia) [reply src/Tests/Microsoft.NET.Sdk.Razor.Tests/BuildIntegrationTest.cs]: I added a new `RazorSdkTest` subclass that these tests build upon. It allows us to override the TFM in the projects when it is copied over. With this, we'll only need to change the TFM there to get things working.

Also updated the TFM to net6.0 for these projects.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs [hunk 15]
```diff
@@ -0,0 +1,39 @@
+using Microsoft.NET.TestFramework.Commands;
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs:1]: License header.

Comment (pranavkm) [reply src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs:1]: You might also not need a bunch of these namespaces.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+        {
+            var projectDirectory = _testAssetsManager
+                .CopyTestAsset(testAsset, callingMethod: callerName, testAssetSubdirectory: subdirectory)
+                .WithSource()
+                .WithProjectChanges(project => 
+                {
+                    var ns = project.Root.Name.Namespace;
+                    var targetFramework = project.Descendants()
+                       .Single(e => e.Name.LocalName == "TargetFramework");
+                    if (targetFramework.Value == "$(DefaultTfm)") {
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs:32]: Formatting.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+using System.Threading;
+using System.Xml.Linq;
+using Microsoft.NET.TestFramework;
+using Xunit.Abstractions;
+
+namespace Microsoft.NET.Sdk.Razor.Tests
+{
+    public abstract class RazorSdkTest : SdkTest
+    {
+        public readonly string DefaultTfm = "net6.0";
```
Comment (pranavkm) [inline src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs:18]: You could stash the tfm of the test project as an assembly attribute and use that here. Here's what that would look like: https://github.com/dotnet/aspnetcore/blob/master/src/Razor/Microsoft.NET.Sdk.Razor/integrationtests/Microsoft.NET.Sdk.Razor.IntegrationTests.csproj#L61-L66

https://github.com/dotnet/aspnetcore/blob/44c0e6651e38079cada6e14ac8978cdb9fd59cc1/src/Shared/MSBuild.Testing/BuildVariables.cs#L14

Of course this is assuming the test tfm gets updated every release.

**Review by pranavkm — APPROVED** [review]
Looks good. I'll do anothert pass through .cshtml and .razor test files and try and strip them down  further. We don't need elaborate apps just enough to verify these compile.

**Review by wli3 — DISMISSED** [review]
hold for more discussion on SDK side

**Comment by wli3** [comment]
this is a build without this change. Windows legs are mostly under 1 hour. However, with this change. It adds 20minutes constantly.

image

**Comment by wli3** [comment]
Compare with release/5.0.2xx -- no other 2 merges. 

image

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs (hunk 17 above)
Comment (captainsafia) [reply src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs]: This is a great idea! It means we can be mostly hands-free when it comes to TFM updates in the repo.

I've updated the code to use this approach.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs (hunk 16 above)
Comment (captainsafia) [reply src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs]: Fixed.

File: src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs (hunk 15 above)
Comment (captainsafia) [reply src/Tests/Microsoft.NET.Sdk.Razor.Tests/RazorSdkTest.cs]: Fixed.

File: src/Assets/TestProjects/RazorAppWithP2PReference/AppWithP2PReference/Views/Shared/Error.cshtml [hunk 18]
```diff
@@ -0,0 +1,22 @@
+﻿@model ErrorViewModel
+@{
+    ViewData["Title"] = "Error";
+}
+
+<h1 class="text-danger">Error.</h1>
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Assets/TestProjects/RazorAppWithP2PReference/AppWithP2PReference/Views/Shared/Error.cshtml]: I ended up removing it.

File: src/Assets/TestProjects/RazorAppWithP2PReference/AppWithP2PReference/Views/Home/Index.cshtml [hunk 19]
```diff
@@ -0,0 +1,108 @@
+﻿@{
+    ViewData["Title"] = "Home Page";
+}
+
+<div id="myCarousel" class="carousel slide" data-ride="carousel" data-interval="6000">
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Assets/TestProjects/RazorAppWithP2PReference/AppWithP2PReference/Views/Home/Index.cshtml]: I trimmed it down here and elsewhere.

**Comment by wli3** [comment]
I don't think reducing the tests are the way to go. Looks like it is still adding time instead of scaling out. Could you work with Helix folks to understanding the reason behind it? Or ask for better machines? Considering SDK repo has multiple team contributing now. Alternatively, you can contribute few low hanging fruits to get the overall time down to 1 hour in Windows, 1:20 in Ubuntu. (Ubuntu is slow still due to I did not parallelize the non helix tests and helix tests due to not enough time)

**Comment by wli3** [comment]
Also I'd like https://github.com/dotnet/aspnetcore/issues/29268#issuecomment-759634279 to finish first. It is better to wrap up one thing first.

**Comment by mkArtakMSFT** [comment]
We've discussed the concerns raised on this migration with @marcpopMSFT and have decided to move forward with these changes.
@captainsafia, going forward you can seek reviews only from @dotnet/aspnet-blazor-eng if your changes scoped to the code our team owns or you need guidance from the SDK team.

**Comment by wli3** [comment]
@mkArtakMSFT I'd like to review the build&test time impact and confirm the concern are addressed before it get merged

**Comment by wli3** [comment]
This is the last PR run 
image

**Comment by wli3** [comment]
I reviewed non Razor specific code (like testframework and overlay logic). It looks good.

**Pull request merged**